feat(trade): display Limit orders promo banner in trading widgets

fairlighteth opened the pull request:
# Add Limit Orders Promotional Banner Across Trading Modes

## Overview
Added a new promotional banner for limit orders that appears across various trading modes (Swap, Limit, TWAP, and Yield). The banner highlights new limit order features and can be dismissed by users, with the dismissal state persisted in local storage.

<img width="508" alt="Screenshot 2025-01-22 at 17 11 28" src="https://github.com/user-attachments/assets/6495ca14-389c-46d9-adb1-958adcaf6fab" />


## Key Changes

### 1. Feature Flag Implementation
- Added `SHOW_LIMIT_ORDERS_PROMO` feature flag to control the banner's visibility globally
- Banner respects both the feature flag and user's dismissal state
- ⚠️ **Important**: Feature flag is currently implemented as a hardcoded constant (`true`), deliberately bypassing the LaunchDarkly feature flag system (from `useFeatureFlags` hook) for initial testing

### 2. Banner Component Updates
- Created `LimitOrdersPromoBanner` pure component with:
  - Bullet points highlighting new limit order features
  - CTA button to navigate to limit orders
  - Dismissible via close button or "Maybe next time" link
  - Conditional rendering of "Maybe next time" link based on current tab
- Added container component to handle:
  - Dismissal state management using `limitOrdersPromoDismissedAtom`
  - Navigation to limit orders page
  - Prop management for child components

### 3. Integration Across Trading Modes
- **Swap Mode**: 
  - Added banner to `TradeWidgetForm`
  - Ensured regular UI is visible when banner is dismissed
  
- **Limit Orders Mode**:
  - Added banner with `isLimitOrdersTab={true}`
  - Hides "Maybe next time" link in this mode
  - Shows regular UI content when dismissed
  
- **TWAP/Advanced Orders Mode**:
  - Added banner with `isLimitOrdersTab={false}`
  - Shows full banner with all elements
  - Maintains existing unlock screen behavior
  
- **Yield Mode**:
  - Added banner with `isLimitOrdersTab={false}`
  - Integrated with existing widget structure
  - Preserves vampire attack banner logic

### 4. UI/UX Improvements
- Banner takes precedence over unlock screens when feature flag is enabled
- Consistent dismissal behavior across all modes
- Smooth transition between banner and regular UI content
- Maintains existing widget functionality when banner is dismissed

### 5. State Management
- Implemented `limitOrdersPromoDismissedAtom` for persistent dismissal state
- Proper state handling for feature flag and dismissal combinations
- Consistent state management across all trading modes

## Testing Notes
- Verify banner appears correctly in all trading modes
- Check dismissal persistence works across page reloads
- Ensure proper navigation when clicking CTA button
- Verify "Maybe next time" link only appears in appropriate contexts
- Test interaction with unlock screens and other UI elements

## Technical Debt
- Removed debug console logs
- Maintained consistent prop passing patterns
- Cleaned up unused imports and variables

## Future Work
- Replace hardcoded `SHOW_LIMIT_ORDERS_PROMO` constant with LaunchDarkly feature flag
- Integrate with `useFeatureFlags` hook for proper feature flag management

anxolin reviewed (approved):
Looks good to me, are these text final. They felt a bit long.  I was thinking sth like:

Limit Order just got better
We heard you! here are some features you've asked for:
- Lock/unlock price for finer control
- Set your price in USD or tokens
- Personalise your interface
- Estimate fill price estimation before placing the order
- More spacious, and enhanced design!

shoom3301 commented on apps/cowswap-frontend/src/pages/AdvancedOrders/index.tsx:
```diff
@@ -24,8 +24,14 @@ import {
-  const { isUnlocked } = useAtomValue(advancedOrdersAtom)
+  const { isUnlocked: isWidgetUnlocked } = useAtomValue(advancedOrdersAtom)
+  const isDismissed = useAtomValue(limitOrdersPromoDismissedAtom)
+  const shouldShowPromo = SHOW_LIMIT_ORDERS_PROMO && !isDismissed
```
This is TWAP module but it uses const that belongs to limit orders, that's not right

shoom3301 commented on apps/cowswap-frontend/src/common/containers/LimitOrdersPromoBanner/index.tsx:
```diff
@@ -0,0 +1,41 @@
+  isLimitOrdersTab?: boolean
+}
+
+export function LimitOrdersPromoBanner({ isLimitOrdersTab = false }: LimitOrdersPromoBannerProps) {
```
Why does it call `LimitOrdersPromoBanner` and it is located in `common` directory?

shoom3301 commented on apps/cowswap-frontend/src/common/containers/LimitOrdersPromoBanner/index.tsx:
```diff
@@ -0,0 +1,41 @@
+  const [isDismissed, setIsDismissed] = useAtom(limitOrdersPromoDismissedAtom)
+  const tradeNavigate = useTradeNavigate()
+
+  const handleDismiss = useCallback(() => {
```
```
const handleDismiss = () => setIsDismissed(true)
```

No need to use `useCallback` since the last React update

fairlighteth replied:
This is what I read on it https://github.com/facebook/react/issues/31913#issuecomment-2568316803

But I will remove it.

shoom3301 commented on apps/cowswap-frontend/src/common/pure/LimitOrdersPromoBanner/index.tsx:
```diff
@@ -0,0 +1,53 @@
+  'Arrange fields to reflect your trading style and preferences.',
+]
+
+export function LimitOrdersPromoBanner({
```
Same question: https://github.com/cowprotocol/cowswap/pull/5325/files#r1926504634

shoom3301 commented on apps/cowswap-frontend/src/modules/limitOrders/containers/LimitOrdersWidget/index.tsx:
```diff
@@ -77,6 +83,9 @@ export function LimitOrdersWidget() {
   const rateInfoParams = useRateInfoParams(inputCurrencyAmount, outputCurrencyAmount)
   const widgetActions = useLimitOrdersWidgetActions()
   const isWrapOrUnwrap = useIsWrapOrUnwrap()
+  const { chainId: _chainId, account: _account } = useWalletInfo()
```
Those variables looks unused

fairlighteth replied:
Addressed and removed some more unused.

shoom3301 reviewed (changes requested):
I think this solution should be different.
The new banner is supposed to be a part of `modules/trade` or `modules/tradeWidgetAddons` and be injected in `TradeWidgetForm` component.
Following this way we can avoid multiple adding of the banner in different places.

elena-zh reviewed (changes requested):
Hey @fairlighteth , great, thank you!
However, some issues: 
1. I cannot check cases related to the 'switched off flag' (and there is no such flag in the launchdarkly)
2. Hooks tab appears to be abandoned with this change
![hoks with banner](https://github.com/user-attachments/assets/5314abd7-5094-4e6e-a4a4-c192a9956dd7)

3. I notice, that when I dismiss (press on later) then navigate to the Limit orders page,  or press on the 'place limit order' button, then refresh the page, the banner may blinks (1 time). 

Might it be possible to fix it?

fairlighteth commented:
Todo once review approved: Use an actual Darkly featureFlag

fairlighteth commented:
Addressed an issue I had found while testing myself:
 
## Changes
- Converted promo banner state from React's `useState` to Jotai's `atomWithStorage` for consistent state management
- Fixed reactivity issues by properly memoizing widget unlock state with `useMemo`
- Simplified promo banner visibility logic into a single `shouldBeVisible` condition
- Removed direct `localStorage` manipulation in favor of Jotai's storage management

## Key Behaviors
- When `SHOW_LIMIT_ORDERS_PROMO` feature flag is enabled:
  - If limit orders promo banner is showing (not dismissed) → all UI components locked
  - If limit orders promo banner is dismissed → all UI components unlocked
- When `SHOW_LIMIT_ORDERS_PROMO` feature flag is disabled:
  - Falls back to regular widget unlock logic

These changes ensure immediate UI updates when dismissing the limit orders promo banner without requiring tab switches (which was the issue at hand).

elena-zh commented:
Hey @fairlighteth , thank you!
1. flag off state tested in #5333 : looks good
2. Hooks: once settings are disabled, not possible to switch the page on --> no longer an issue
3. widget: new global banner is not displayed there
4. **TWAP page: when I press on the 'later' button, the screen is unlocked, but another banner - unlock TWAP appears, however, it should not**
![image](https://github.com/user-attachments/assets/d8585bcf-44f9-44b6-95fa-ae3e7a5e864b)

Could you please fix it?

shoom3301 commented on apps/cowswap-frontend/src/common/state/limitOrdersPromoAtom.ts:
```diff
@@ -0,0 +1,3 @@
+import { atomWithStorage } from 'jotai/utils'
+
+export const limitOrdersPromoDismissedAtom = atomWithStorage<boolean>('limitOrdersPromo_dismissed', false)
```
This file seems to be unused

shoom3301 commented on apps/cowswap-frontend/src/modules/trade/containers/TradeWidget/index.tsx:
```diff
@@ -14,7 +19,40 @@ export function TradeWidget(props: TradeWidgetProps) {
+  const { isVisible, onDismiss, isLimitOrdersTab } = useLimitOrdersPromoBanner()
+  const navigate = useNavigate()
+
+  const handleCtaClick = () => {
```
I think it makes sense to add a smart component as a wrapper around `LimitOrdersPromoBanner` which will define `handleCtaClick` inside

shoom3301 replied:
I see that you need to decide if the banner should be shown and display another content if not.
You can pass the fallback content as:
```
<LimitOrdersPromoBanner>
   {slots.topContent}
</LimitOrdersPromoBanner>
```

And inside of the compontent take a decision based on `isVisible` value

fairlighteth replied:
Thanks for checking this. Refactored this now.

shoom3301 commented on apps/cowswap-frontend/src/modules/advancedOrders/hooks/useIsWidgetUnlocked.ts:
```diff
@@ -0,0 +1,9 @@
+
+import { AdvancedOrdersRawState } from '../state/advancedOrdersAtom'
+
+export function useIsWidgetUnlocked(): boolean {
```
This file is also not in use

fairlighteth replied:
Addressed

shoom3301 commented on apps/cowswap-frontend/src/modules/limitOrders/containers/LimitOrdersWidget/index.tsx:
```diff
@@ -216,7 +220,13 @@ const LimitOrders = React.memo((props: LimitOrdersProps) => {
       )
     },
-    outerContent: <>{isUnlocked && <InfoBanner />}</>,
+    outerContent: (
```
I think you can also handle this logic inside of `TradeWidget` component like you have done with `topContent` and `lockScreen`

shoom3301 commented on apps/cowswap-frontend/src/modules/trade/containers/TradeWidget/index.tsx:
```diff
@@ -14,7 +17,21 @@ export function TradeWidget(props: TradeWidgetProps) {
+    topContent: (
+      <>
+        <LimitOrdersPromoBannerWrapper />
+        {slots.topContent}
```
topContent was showing only when `isVisible === false` is it not true anymore?

shoom3301 commented on apps/cowswap-frontend/src/modules/trade/containers/TradeWidget/index.tsx:
```diff
@@ -14,7 +17,21 @@ export function TradeWidget(props: TradeWidgetProps) {
+      </>
+    ),
+    // Only use banner as lockScreen when it's visible, otherwise use original lockScreen
+    lockScreen: isVisible ? <LimitOrdersPromoBannerWrapper /> : slots.lockScreen,
```
I meant you can do:
```
<LimitOrdersPromoBannerWrapper>{slots.lockScreen}</LimitOrdersPromoBannerWrapper>
```

And inside of `LimitOrdersPromoBannerWrapper`:

```
function LimitOrdersPromoBannerWrapper({children}) {
const { isVisible } = useLimitOrdersPromoBanner()

if (!isVisible) return children
}
```

Anyway, since we don't have too much time, could you add TODO for that please? Not necessary to fix right now

shoom3301 commented on apps/cowswap-frontend/src/modules/trade/containers/TradeWidget/index.tsx:
```diff
@@ -4,6 +4,7 @@ import { TradeWidgetModals } from './TradeWidgetModals'
 import { TradeWidgetUpdaters } from './TradeWidgetUpdaters'
 import { TradeWidgetProps } from './types'
 
+import { useTradeFlowContext } from '../../../limitOrders/hooks/useTradeFlowContext'
```
`trade` module cannot depend on `limitOrders` module

shoom3301 replied:
Keep it in limit orders widget

shoom3301 commented:
@elena-zh I know, there is a new `skipLockScreen` key in URL, just ignore it...

elena-zh commented:
I see an 'old' unlock screen on Limit orders page when I close the new one (the flag is ON)
![image](https://github.com/user-attachments/assets/16f83b92-ce95-48da-94e8-c8555cf42519)

alfetopito commented:
> I see an 'old' unlock screen on Limit orders page when I close the new one (the flag is ON) ![image](https://private-user-images.githubusercontent.com/70885163/406486794-16f83b92-ce95-48da-94e8-c8555cf42519.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc3MzIzMDAsIm5iZiI6MTczNzczMjAwMCwicGF0aCI6Ii83MDg4NTE2My80MDY0ODY3OTQtMTZmODNiOTItY2U5NS00OGRhLTk0ZTgtYzg1NTVjZjQyNTE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTI0VDE1MjAwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNkYTA2Y2ZkMjgzODExMjQ3MTZmN2ZmYmQ2MzQ1YWNmNzkzMGZkMGRjYTZjOWVmMDM0OWFmOTBkN2YwZTBkNzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.RFB3z4nwy15nqdHz_MoJu8uVWZeRf1IC5ObUHi1IO_8)
> 
> Reproducible when:
> 
>     1. I'm no TWAP
> 
>     2. I press on the 'Maybe next time'
> 
>     3. Then navigate to the limit order page

If that's the only problem left, let's merge it and address it in another PR
